// h-vue-tabs/dist/es.js
import { inject as p, ref as f, watch as m, onMounted as y, openBlock as s, createElementBlock as u, createElementVNode as E, Fragment as x, renderList as w, normalizeStyle as h, toDisplayString as S } from "vue";
/*!
  * vue-router v4.2.4
  * (c) 2023 Eduardo San Martin Morote
  * @license MIT
  */
var d;
(function(e) {
  e.pop = "pop", e.push = "push";
})(d || (d = {}));
var i;
(function(e) {
  e.back = "back", e.forward = "forward", e.unknown = "";
})(i || (i = {}));
Symbol(process.env.NODE_ENV !== "production" ? "navigation failure" : "");
var b;
(function(e) {
  e[e.aborted = 4] = "aborted", e[e.cancelled = 8] = "cancelled", e[e.duplicated = 16] = "duplicated";
})(b || (b = {}));
Symbol(process.env.NODE_ENV !== "production" ? "router view location matched" : "");
Symbol(process.env.NODE_ENV !== "production" ? "router view depth" : "");
const T = Symbol(process.env.NODE_ENV !== "production" ? "router" : ""), N = Symbol(process.env.NODE_ENV !== "production" ? "route location" : "");
Symbol(process.env.NODE_ENV !== "production" ? "router view location" : "");
function k() {
  return p(T);
}
function C() {
  return p(N);
}
const O = (e, o) => {
  const t = e.__vccOpts || e;
  for (const [r, a] of o)
    t[r] = a;
  return t;
}, V = { class: "tab_wrapper" }, I = { class: "tab_list" }, D = ["onClick"], g = {
  __name: "Tab",
  props: {
    tabs: {
      type: String,
      default: () => ["tab1", "tab2"],
      required: !0
    },
    activeTabColor: {
      type: String,
      default: "#000000"
    },
    activeTabIndex: {
      type: Number,
      default: 0
    }
  },
  setup(e) {
    const o = e, t = f(), r = k();
    C(), m(
      () => o.activeTabIndex,
      () => {
        t.value = o.tabs.findIndex(
          (c) => c === o.activeTabIndex
        );
      }
    );
    const a = (c, l) => {
      t.value = c.toLowerCase(), r.push({ query: { activeTab: t.value } });
    };
    return y(() => {
      const c = o.tabs.find((l, n) => n === o.activeTabIndex);
      t.value = c.toLowerCase(), console.log("activeTab", t.value);
    }), (c, l) => (s(), u("div", V, [
      E("ul", I, [
        (s(!0), u(x, null, w(e.tabs, (n, v) => (s(), u("li", {
          style: h(
            (n == null ? void 0 : n.toLowerCase()) === t.value && `color:${e.activeTabColor}; border-bottom: 4px solid ${e.activeTabColor}; font-weight: 600`
          ),
          onClick: (R) => a(n),
          key: v,
          class: "tab_data"
        }, S(n), 13, D))), 128))
      ])
    ]));
  }
}, _ = /* @__PURE__ */ O(g, [["__scopeId", "data-v-dfbae85e"]]);
function L(e) {
  e.component("Tab", _);
}
const q = {
  install: L,
  version: "__VERSION__",
  Tab: _
};
export {
  q as default
};
